File: backend/app/schemas/Suggestion.py
from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


SuggestionResourceType = Literal["LESSON", "CLASSWORK"]
SuggestionPriority = Literal["LOW", "NORMAL", "HIGH", "URGENT"]
SuggestionStatus = Literal["DRAFT", "ACTIVE", "COMPLETED", "DISMISSED", "ARCHIVED"]


class ManualSuggestionCreate(BaseModel):
    student_id: UUID
    subject_id: int
    resource_type: SuggestionResourceType
    title: str = Field(min_length=1, max_length=255)
    description: Optional[str] = None
    priority: SuggestionPriority = "NORMAL"
    lesson_id: Optional[int] = None
    classwork_assignment_id: Optional[int] = None


class SuggestionResourceSummary(BaseModel):
    resource_type: SuggestionResourceType
    title: str
    subject_id: int
    is_available: bool
    unavailable_reason: Optional[str] = None
    lesson_id: Optional[int] = None
    classwork_id: Optional[int] = None
    classwork_assignment_id: Optional[int] = None
    classwork_type: Optional[str] = None
    class_id: Optional[int] = None


class SuggestionClassworkResponse(BaseModel):
    suggestion_classwork_id: int
    classwork_assignment_id: int
    is_completed: bool
    completed_at: Optional[datetime] = None
    score_before: Optional[float] = None
    score_after: Optional[float] = None


class SuggestionResponse(BaseModel):
    student_suggestion_id: int
    suggestion_type: str
    resource_type: SuggestionResourceType
    title: str
    description: Optional[str] = None
    priority: SuggestionPriority
    status: SuggestionStatus
    is_viewed: bool
    viewed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    student_id: UUID
    subject_id: int
    lesson_id: Optional[int] = None
    created_by_staff_id: Optional[str] = None
    resource: SuggestionResourceSummary
    classwork_link: Optional[SuggestionClassworkResponse] = None
    source_metrics: Optional[dict] = None


class RecommendationDraftRequest(BaseModel):
    class_id: int
    subject_id: int
    low_score_threshold: float = Field(default=75, ge=0, le=100)
    limit: int = Field(default=20, ge=1, le=100)


class SuggestionListResponse(BaseModel):
    suggestions: list[SuggestionResponse] = Field(default_factory=list)

File: frontend/src/types/suggestion.ts
export type SuggestionResourceType = "LESSON" | "CLASSWORK";
export type SuggestionPriority = "LOW" | "NORMAL" | "HIGH" | "URGENT";
export type SuggestionStatus = "DRAFT" | "ACTIVE" | "COMPLETED" | "DISMISSED" | "ARCHIVED";

export type SuggestionResourceSummary = {
  resource_type: SuggestionResourceType;
  title: string;
  subject_id: number;
  is_available: boolean;
  unavailable_reason?: string | null;
  lesson_id?: number | null;
  classwork_id?: number | null;
  classwork_assignment_id?: number | null;
  classwork_type?: string | null;
  class_id?: number | null;
};

export type SuggestionClassworkResponse = {
  suggestion_classwork_id: number;
  classwork_assignment_id: number;
  is_completed: boolean;
  completed_at?: string | null;
  score_before?: number | null;
  score_after?: number | null;
};

export type SuggestionResponse = {
  student_suggestion_id: number;
  suggestion_type: string;
  resource_type: SuggestionResourceType;
  title: string;
  description?: string | null;
  priority: SuggestionPriority;
  status: SuggestionStatus;
  is_viewed: boolean;
  viewed_at?: string | null;
  created_at?: string | null;
  student_id: string;
  subject_id: number;
  lesson_id?: number | null;
  created_by_staff_id?: string | null;
  resource: SuggestionResourceSummary;
  classwork_link?: SuggestionClassworkResponse | null;
  source_metrics?: Record<string, unknown> | null;
};

export type SuggestionListResponse = {
  suggestions: SuggestionResponse[];
};

export type ManualSuggestionCreate = {
  student_id: string;
  subject_id: number;
  resource_type: SuggestionResourceType;
  title: string;
  description?: string | null;
  priority: SuggestionPriority;
  lesson_id?: number | null;
  classwork_assignment_id?: number | null;
};

export type RecommendationDraftRequest = {
  class_id: number;
  subject_id: number;
  low_score_threshold?: number;
  limit?: number;
};
